software/controller/lib/hal/i2c.cpp

```diff
@@ -0,0 +1,497 @@
+            // must detect that the next write data actually starts at index 0
+            // to properly free the end of the buffer as well
+            write_buffer_start_ = 0;
+          }
+        }
+        // start transfer to initiate the next request (if there is one)
+        StartTransfer();
+      }
+    }
+    if (NackDetected()) {
+      // clear the nack
+#if defined(BARE_STM32)
+      i2c_->intClr = 0x10;
+#endif
+      // the peripheral is non-responsive --> start the request anew
+      next_data_ = reinterpret_cast<uint8_t *>(last_request_.data);
+      remaining_size_ = last_request_.size;
+      StartTransfer();
+    }
+  }
+}
+
+void I2CChannel::I2CErrorHandler() {
+  // i2c error --> clear all error flags (except those that are SMBus only)
+#if defined(BARE_STM32)
+  i2c_->intClr = 0x720;
+#endif
+  // and restart the request up to error_retry_ times
+  if (--error_retry_ > 0) {
+    next_data_ = reinterpret_cast<uint8_t *>(last_request_.data);
+    remaining_size_ = last_request_.size;
+  } else {
+    // skip this request and go to next one;
+    remaining_size_ = 0;
+  }
+  StartTransfer();
+}
+
+#if defined(BARE_STM32)
+void I2CChannel::DMAIntHandler(DMA_Chan chan) {
+  if (dma_enable_ && transfer_in_progress_) {
+    if (DMA_IntStatus(dma_, chan, DmaInterrupt::XFER_COMPLETE)) {
+      if (remaining_size_ > 255) {
+        // decrement remaining size by 255 (the size of the DMA transfer)
+        remaining_size_ = static_cast<uint16_t>(remaining_size_ - 255);
+      } else {
+        // indicate the last request as processed
+        *last_request_.processed = true;
+        transfer_in_progress_ = false;
+        i2c_->ctrl2.stop = 1;
+        remaining_size_ = 0;
+        dma_->channel[static_cast<uint8_t>(chan)].config.enable = 0;
+        if (last_request_.read_write == I2CExchangeDir::kWrite) {
+          // free the part of the write buffer that was dedicated to this
+          // request
+          write_buffer_start_ += last_request_.size;
+          if (write_buffer_start_ >= wrapping_index_) {
+            // We don't allow data in the same request to wrap around, so we
+            // must detect that the next write data actually starts at index 0
+            // to properly free the end of the buffer as well
+            write_buffer_start_ = 0;
+          }
+        }
+      }
+    } else if (DMA_IntStatus(dma_, chan, DmaInterrupt::XFER_ERR)) {
+      // we are dealing with an error --> reset transfer (up to kMaxRetries
+      // times)
+      if (--error_retry_ > 0) {
+        next_data_ = reinterpret_cast<uint8_t *>(last_request_.data);
+        remaining_size_ = last_request_.size;
+      } else {
+        // skip this request and go to next request;
+        remaining_size_ = 0;
+      }
+    }
+    // clear all interrupts and (re-)start the current or next transfer
+    DMA_ClearInt(dma_, chan, DmaInterrupt::GLOBAL);
+    StartTransfer();
+  }
+}
```

Comment:
I will do my best to break this out into smaller functions, but I invite you to read figures 369 though 373 of the [RM], which should make this less voodoo.
The only major difference between my implementation and those flowcharts is the use of interrupts instead of busy wait.